src/classes/TestResult.js

```diff
@@ -148,7 +172,9 @@ export default class TestResult extends BubblingEventTarget {
 					this.skip();
 				}
 				else {
-					this.run();
+					this.setup()
+						.then(() => this.run())
+						.then(() => this.teardown());
```

Comment:
```suggestion
						.finally(() => this.teardown());
```

Reply:
Not a blocker, but we eventually need to address what happens when these error as a test failure, just like we did with map and check (perhaps time for an abstraction).